sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/AmqpMetricsProvider.java

```diff
@@ -0,0 +1,397 @@
+    public void recordReceivedMessage(Message message) {
+        if (!isEnabled || !receivedLag.isEnabled()
+            || message == null
+            || message.getMessageAnnotations() == null
+            || message.getBody() == null) {
+            return;
+        }
+
+        Map<Symbol, Object> properties = message.getMessageAnnotations().getValue();
+        Object enqueuedTimeDate = properties != null ? properties.get(ENQUEUED_TIME_ANNOTATION) : null;
+        if (enqueuedTimeDate instanceof Date) {
+            Instant enqueuedTime = ((Date) enqueuedTimeDate).toInstant();
+            long deltaMs = Instant.now().toEpochMilli() - enqueuedTime.toEpochMilli();
+            if (deltaMs < 0) {
+                deltaMs = 0;
+            }
+            receivedLag.record(deltaMs / 1000d, commonAttributes, Context.NONE);
+        } else {
+            LOGGER.verbose("Received message has unexpected `x-opt-enqueued-time` annotation value - `{}`. Ignoring it.", enqueuedTimeDate);
+        }
+    }
+
+    /**
+     * Records that credits were added to link
+     */
+    public void recordAddCredits(int credits) {
+        if (isEnabled && addCredits.isEnabled()) {
+            addCredits.add(credits, commonAttributes, Context.NONE);
+        }
+    }
+
+    /**
+     * Records link error. Noop if condition is null (no error).
+     */
+    public void recordHandlerError(ErrorSource source, ErrorCondition condition) {
+        if (isEnabled && condition != null && condition.getCondition() != null) {
+            TelemetryAttributes attributes = amqpErrorAttributeCache.getOrCreate(condition.getCondition().toString());
+            switch (source) {
+                case LINK:
+                    if (linkErrors.isEnabled()) {
+                        linkErrors.add(1, attributes, Context.NONE);
+                    }
+                    break;
+                case SESSION:
+                    if (sessionErrors.isEnabled()) {
+                        sessionErrors.add(1, attributes, Context.NONE);
+                    }
+                    break;
+                case TRANSPORT:
+                    if (transportErrors.isEnabled()) {
+                        transportErrors.add(1, attributes, Context.NONE);
+                    }
+                    break;
+                default:
+                    LOGGER.verbose("Unexpected error source: {}", source);
+            }
+        }
+    }
+
+    private TelemetryAttributes getDeliveryStateAttribute(DeliveryState.DeliveryStateType state) {
+        int ind = state == null ? DELIVERY_STATES_COUNT - 1 : state.ordinal();
+        TelemetryAttributes attrs = sendAttributeCache[ind];
+        if (attrs != null) {
+            return attrs;
+        }
+
+        return createDeliveryStateAttribute(state, ind);
+    }
+
+    private TelemetryAttributes getResponseCodeAttributes(AmqpResponseCode code, String operation) {
+        int ind = code == null ? RESPONSE_CODES_COUNT - 1 : code.ordinal();
```

Comment:
I am curious why we have the `- 1` here and then `+ 1` when defining RESPONSE_CODES_COUNT.  A comment where RESPONSE_CODES_COUNT could help in the future if someone is checking this out.

Reply:
thanks for the feedback, I added comments in the code!